File: product-of-array-except-self/product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] leftProd = new int[nums.length];
        int[] rightProd = new int[nums.length];
        int[] ans = new int[nums.length];

        leftProd[0] = 1;
        rightProd[nums.length - 1] = 1;
        for (int i = 1; i < nums.length; i++) {
            leftProd[i] = nums[i - 1] * leftProd[i - 1]; 
        }

        for (int i = nums.length - 2; i >= 0; i--) {
            rightProd[i] = nums[i + 1] * rightProd[i + 1]; 
        }

        for (int i = 0; i < nums.length; i++) {
            ans[i] = leftProd[i] * rightProd[i];
        }

        return ans;
    }
}

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    # def productExceptSelf(self, nums: List[int]) -> List[int]:
    #     leftProdArr = [1] * len(nums)
    #     rightProdArr = [1] * len(nums)

    #     prod = 1
    #     for i in range(1, len(nums)):
    #         leftProdArr[i] = prod * nums[i - 1]
    #         prod = leftProdArr[i]

    #     prod = 1
    #     for j in range(len(nums) - 2, -1, -1):
    #         rightProdArr[j] = prod * nums[j + 1]
    #         prod = rightProdArr[j]

    #     output = [0] * len(nums)

    #     for i in range(0, len(nums)):
    #         output[i] = leftProdArr[i] * rightProdArr[i]

    #     return output    

    def productExceptSelf(self, nums: List[int]) -> List[int]:
        output = [1] * len(nums)

        leftProd = 1
        for i in range(len(nums)):
            output[i] *= leftProd
            leftProd *= nums[i] 

        rightProd = 1
        for j in range(len(nums) - 1, -1, -1):
            output[j] *= rightProd
            rightProd *= nums[j]


        return output    
        